File: cognito/scripts/test_api.py
#!/usr/bin/env python3
"""End-to-end API testing script for the Cognito Auth System.

Usage:
    python scripts/test_api.py --api-url https://abc123.execute-api.us-east-1.amazonaws.com/dev
    python scripts/test_api.py --api-url https://abc123.execute-api.us-east-1.amazonaws.com/dev \
                               --username admin@example.com --password 'P@ssw0rd!'

The script exercises:
    1. Login flow (POST /auth/authorize)
    2. Token refresh (POST /auth/refresh)
    3. Protected endpoints (GET/POST/DELETE /events)
    4. Authorization failure scenarios
"""

import argparse
import json
import sys
from typing import Any

import requests

# ---------------------------------------------------------------------------
# Colours for terminal output
# ---------------------------------------------------------------------------
GREEN = "\033[92m"
RED = "\033[91m"
YELLOW = "\033[93m"
RESET = "\033[0m"


def _ok(msg: str) -> None:
    print(f"  {GREEN}✓ PASS{RESET}  {msg}")


def _fail(msg: str) -> None:
    print(f"  {RED}✗ FAIL{RESET}  {msg}")


def _info(msg: str) -> None:
    print(f"  {YELLOW}ℹ INFO{RESET}  {msg}")


# ---------------------------------------------------------------------------
# Test helpers
# ---------------------------------------------------------------------------
def login(api_url: str, username: str, password: str) -> dict | None:
    """Authenticate and return the token payload, or None on failure."""
    url = f"{api_url.rstrip('/')}/auth/authorize"
    print(f"\n{'='*60}")
    print(f"POST {url}")
    print(f"{'='*60}")

    resp = requests.post(url, json={"username": username, "password": password}, timeout=15)
    body = resp.json()

    if resp.status_code == 200:
        _ok(f"Login successful (HTTP {resp.status_code})")
        _info(f"AccessToken: {body['tokens']['AccessToken'][:40]}...")
        return body["tokens"]
    else:
        _fail(f"Login failed (HTTP {resp.status_code}): {body.get('message')}")
        return None


def refresh_tokens(api_url: str, refresh_token: str) -> dict | None:
    """Refresh tokens and return the new token payload."""
    url = f"{api_url.rstrip('/')}/auth/refresh"
    print(f"\n{'='*60}")
    print(f"POST {url}")
    print(f"{'='*60}")

    resp = requests.post(url, json={"refresh_token": refresh_token}, timeout=15)
    body = resp.json()

    if resp.status_code == 200:
        _ok(f"Token refresh successful (HTTP {resp.status_code})")
        return body["tokens"]
    else:
        _fail(f"Token refresh failed (HTTP {resp.status_code}): {body.get('message')}")
        return None


def test_list_events(api_url: str, access_token: str) -> None:
    """GET /events with a valid token."""
    url = f"{api_url.rstrip('/')}/events"
    print(f"\n{'='*60}")
    print(f"GET {url}")
    print(f"{'='*60}")

    resp = requests.get(url, headers=_auth_header(access_token), timeout=15)
    body = resp.json()

    if resp.status_code == 200:
        _ok(f"List events (HTTP {resp.status_code}) – {body.get('count', 0)} event(s)")
    else:
        _fail(f"List events failed (HTTP {resp.status_code}): {body}")


def test_create_event(api_url: str, access_token: str) -> str | None:
    """POST /events – create a test event; return eventId on success."""
    url = f"{api_url.rstrip('/')}/events"
    print(f"\n{'='*60}")
    print(f"POST {url}")
    print(f"{'='*60}")

    payload = {
        "title": "Test Deployment v2.1",
        "description": "Automated test event created by test_api.py",
        "visibility": "public",
    }
    resp = requests.post(url, headers=_auth_header(access_token), json=payload, timeout=15)
    body = resp.json()

    if resp.status_code == 201:
        event_id = body.get("event", {}).get("eventId", "")
        _ok(f"Event created (HTTP {resp.status_code}) – eventId={event_id}")
        return event_id
    else:
        _fail(f"Create event failed (HTTP {resp.status_code}): {body}")
        return None


def test_get_event(api_url: str, access_token: str, event_id: str) -> None:
    """GET /events/{eventId}."""
    url = f"{api_url.rstrip('/')}/events/{event_id}"
    print(f"\n{'='*60}")
    print(f"GET {url}")
    print(f"{'='*60}")

    resp = requests.get(url, headers=_auth_header(access_token), timeout=15)
    body = resp.json()

    if resp.status_code == 200:
        _ok(f"Get event (HTTP {resp.status_code}) – title={body.get('event', {}).get('title')}")
    else:
        _fail(f"Get event failed (HTTP {resp.status_code}): {body}")


def test_delete_event(api_url: str, access_token: str, event_id: str) -> None:
    """DELETE /events/{eventId}."""
    url = f"{api_url.rstrip('/')}/events/{event_id}"
    print(f"\n{'='*60}")
    print(f"DELETE {url}")
    print(f"{'='*60}")

    resp = requests.delete(url, headers=_auth_header(access_token), timeout=15)
    body = resp.json()

    if resp.status_code == 200:
        _ok(f"Event deleted (HTTP {resp.status_code})")
    elif resp.status_code == 403:
        _info(f"Delete denied (HTTP 403) – expected if user is not admin: {body.get('message')}")
    else:
        _fail(f"Delete event failed (HTTP {resp.status_code}): {body}")


def test_unauthorized_access(api_url: str) -> None:
    """Call a protected endpoint without a token – expect 401."""
    url = f"{api_url.rstrip('/')}/events"
    print(f"\n{'='*60}")
    print(f"GET {url}  (no Authorization header)")
    print(f"{'='*60}")

    resp = requests.get(url, timeout=15)

    if resp.status_code == 401:
        _ok(f"Correctly rejected (HTTP 401)")
    else:
        _fail(f"Expected 401 but got HTTP {resp.status_code}")


def test_invalid_token(api_url: str) -> None:
    """Call a protected endpoint with a garbage token – expect 401."""
    url = f"{api_url.rstrip('/')}/events"
    print(f"\n{'='*60}")
    print(f"GET {url}  (invalid token)")
    print(f"{'='*60}")

    resp = requests.get(url, headers={"Authorization": "Bearer invalid.token.here"}, timeout=15)

    if resp.status_code in (401, 403):
        _ok(f"Correctly rejected (HTTP {resp.status_code})")
    else:
        _fail(f"Expected 401/403 but got HTTP {resp.status_code}")


# ---------------------------------------------------------------------------
# Utilities
# ---------------------------------------------------------------------------
def _auth_header(token: str) -> dict[str, str]:
    return {"Authorization": f"Bearer {token}"}


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
def main() -> None:
    parser = argparse.ArgumentParser(description="End-to-end API tests")
    parser.add_argument("--api-url", required=True, help="Base API Gateway URL")
    parser.add_argument("--username", required=True, help="Cognito username")
    parser.add_argument("--password", required=True, help="Cognito password")
    args = parser.parse_args()

    print("\n" + "=" * 60)
    print("  Cognito Auth System – End-to-End Tests")
    print("=" * 60)

    # 1. Login
    tokens = login(args.api_url, args.username, args.password)
    if not tokens:
        print("\nAborting – login failed.")
        sys.exit(1)

    access_token: str = tokens["AccessToken"]

    # 2. Refresh
    if tokens.get("RefreshToken"):
        new_tokens = refresh_tokens(args.api_url, tokens["RefreshToken"])
        if new_tokens and new_tokens.get("AccessToken"):
            access_token = new_tokens["AccessToken"]
            _info("Using refreshed AccessToken for remaining tests")

    # 3. List events
    test_list_events(args.api_url, access_token)

    # 4. Create event
    event_id = test_create_event(args.api_url, access_token)

    # 5. Get single event
    if event_id:
        test_get_event(args.api_url, access_token, event_id)

    # 6. Delete event
    if event_id:
        test_delete_event(args.api_url, access_token, event_id)

    # 7. Unauthorized access (no token)
    test_unauthorized_access(args.api_url)

    # 8. Invalid token
    test_invalid_token(args.api_url)

    print(f"\n{'='*60}")
    print("  Tests complete")
    print(f"{'='*60}\n")


if __name__ == "__main__":
    main()

File: scripts/test_api.ts
/**
 * End-to-end API testing script for the Cognito Auth System.
 *
 * Usage:
 *   npx ts-node scripts/test_api.ts --api-url https://abc123.execute-api.us-east-1.amazonaws.com/dev \
 *                                    --username admin@example.com --password 'P@ssw0rd!'
 *
 * The script exercises:
 *   1. Login flow (POST /auth/authorize)
 *   2. Token refresh (POST /auth/refresh)
 *   3. Protected endpoints (GET/POST/DELETE /events)
 *   4. Authorization failure scenarios
 */

// ---------------------------------------------------------------------------
// Colours for terminal output
// ---------------------------------------------------------------------------
const GREEN = "\x1b[92m";
const RED = "\x1b[91m";
const YELLOW = "\x1b[93m";
const RESET = "\x1b[0m";

function ok(msg: string): void {
  console.log(`  ${GREEN}✓ PASS${RESET}  ${msg}`);
}

function fail(msg: string): void {
  console.log(`  ${RED}✗ FAIL${RESET}  ${msg}`);
}

function info(msg: string): void {
  console.log(`  ${YELLOW}ℹ INFO${RESET}  ${msg}`);
}

// ---------------------------------------------------------------------------
// HTTP helper (uses Node.js built-in fetch, available in Node 18+)
// ---------------------------------------------------------------------------
interface FetchResult {
  status: number;
  body: Record<string, unknown>;
}

async function request(
  url: string,
  options: RequestInit = {}
): Promise<FetchResult> {
  const resp = await fetch(url, {
    ...options,
    headers: {
      "Content-Type": "application/json",
      ...(options.headers as Record<string, string>),
    },
    signal: AbortSignal.timeout(15_000),
  });
  const body = (await resp.json()) as Record<string, unknown>;
  return { status: resp.status, body };
}

function authHeader(token: string): Record<string, string> {
  return { Authorization: `Bearer ${token}` };
}

// ---------------------------------------------------------------------------
// Test helpers
// ---------------------------------------------------------------------------

interface Tokens {
  AccessToken?: string;
  IdToken?: string;
  RefreshToken?: string;
  ExpiresIn?: number;
  TokenType?: string;
}

async function login(
  apiUrl: string,
  username: string,
  password: string
): Promise<Tokens | null> {
  const url = `${apiUrl.replace(/\/$/, "")}/auth/authorize`;
  console.log(`\n${"=".repeat(60)}`);
  console.log(`POST ${url}`);
  console.log(`${"=".repeat(60)}`);

  const { status, body } = await request(url, {
    method: "POST",
    body: JSON.stringify({ username, password }),
  });

  if (status === 200) {
    const tokens = body.tokens as Tokens;
    ok(`Login successful (HTTP ${status})`);
    info(`AccessToken: ${tokens.AccessToken?.substring(0, 40)}...`);
    return tokens;
  } else {
    fail(`Login failed (HTTP ${status}): ${(body as Record<string, unknown>).message}`);
    return null;
  }
}

async function refreshTokens(
  apiUrl: string,
  refreshToken: string
): Promise<Tokens | null> {
  const url = `${apiUrl.replace(/\/$/, "")}/auth/refresh`;
  console.log(`\n${"=".repeat(60)}`);
  console.log(`POST ${url}`);
  console.log(`${"=".repeat(60)}`);

  const { status, body } = await request(url, {
    method: "POST",
    body: JSON.stringify({ refresh_token: refreshToken }),
  });

  if (status === 200) {
    ok(`Token refresh successful (HTTP ${status})`);
    return body.tokens as Tokens;
  } else {
    fail(`Token refresh failed (HTTP ${status}): ${(body as Record<string, unknown>).message}`);
    return null;
  }
}

async function testListEvents(
  apiUrl: string,
  accessToken: string
): Promise<void> {
  const url = `${apiUrl.replace(/\/$/, "")}/events`;
  console.log(`\n${"=".repeat(60)}`);
  console.log(`GET ${url}`);
  console.log(`${"=".repeat(60)}`);

  const { status, body } = await request(url, {
    headers: authHeader(accessToken),
  });

  if (status === 200) {
    ok(`List events (HTTP ${status}) – ${body.count ?? 0} event(s)`);
  } else {
    fail(`List events failed (HTTP ${status}): ${JSON.stringify(body)}`);
  }
}

async function testCreateEvent(
  apiUrl: string,
  accessToken: string
): Promise<string | null> {
  const url = `${apiUrl.replace(/\/$/, "")}/events`;
  console.log(`\n${"=".repeat(60)}`);
  console.log(`POST ${url}`);
  console.log(`${"=".repeat(60)}`);

  const payload = {
    title: "Test Deployment v2.1",
    description: "Automated test event created by test_api.ts",
    visibility: "public",
  };

  const { status, body } = await request(url, {
    method: "POST",
    headers: authHeader(accessToken),
    body: JSON.stringify(payload),
  });

  if (status === 201) {
    const eventData = body.event as Record<string, unknown>;
    const eventId = eventData?.eventId as string;
    ok(`Event created (HTTP ${status}) – eventId=${eventId}`);
    return eventId;
  } else {
    fail(`Create event failed (HTTP ${status}): ${JSON.stringify(body)}`);
    return null;
  }
}

async function testGetEvent(
  apiUrl: string,
  accessToken: string,
  eventId: string
): Promise<void> {
  const url = `${apiUrl.replace(/\/$/, "")}/events/${eventId}`;
  console.log(`\n${"=".repeat(60)}`);
  console.log(`GET ${url}`);
  console.log(`${"=".repeat(60)}`);

  const { status, body } = await request(url, {
    headers: authHeader(accessToken),
  });

  if (status === 200) {
    const eventData = body.event as Record<string, unknown>;
    ok(`Get event (HTTP ${status}) – title=${eventData?.title}`);
  } else {
    fail(`Get event failed (HTTP ${status}): ${JSON.stringify(body)}`);
  }
}

async function testDeleteEvent(
  apiUrl: string,
  accessToken: string,
  eventId: string
): Promise<void> {
  const url = `${apiUrl.replace(/\/$/, "")}/events/${eventId}`;
  console.log(`\n${"=".repeat(60)}`);
  console.log(`DELETE ${url}`);
  console.log(`${"=".repeat(60)}`);

  const { status, body } = await request(url, {
    method: "DELETE",
    headers: authHeader(accessToken),
  });

  if (status === 200) {
    ok(`Event deleted (HTTP ${status})`);
  } else if (status === 403) {
    info(
      `Delete denied (HTTP 403) – expected if user is not admin: ${(body as Record<string, unknown>).message}`
    );
  } else {
    fail(`Delete event failed (HTTP ${status}): ${JSON.stringify(body)}`);
  }
}

async function testUnauthorizedAccess(apiUrl: string): Promise<void> {
  const url = `${apiUrl.replace(/\/$/, "")}/events`;
  console.log(`\n${"=".repeat(60)}`);
  console.log(`GET ${url}  (no Authorization header)`);
  console.log(`${"=".repeat(60)}`);

  const { status } = await request(url);

  if (status === 401) {
    ok(`Correctly rejected (HTTP 401)`);
  } else {
    fail(`Expected 401 but got HTTP ${status}`);
  }
}

async function testInvalidToken(apiUrl: string): Promise<void> {
  const url = `${apiUrl.replace(/\/$/, "")}/events`;
  console.log(`\n${"=".repeat(60)}`);
  console.log(`GET ${url}  (invalid token)`);
  console.log(`${"=".repeat(60)}`);

  const { status } = await request(url, {
    headers: { Authorization: "Bearer invalid.token.here" },
  });

  if (status === 401 || status === 403) {
    ok(`Correctly rejected (HTTP ${status})`);
  } else {
    fail(`Expected 401/403 but got HTTP ${status}`);
  }
}

// ---------------------------------------------------------------------------
// CLI argument parsing
// ---------------------------------------------------------------------------
function parseArgs(): { apiUrl: string; username: string; password: string } {
  const args = process.argv.slice(2);
  let apiUrl = "";
  let username = "";
  let password = "";

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--api-url" && args[i + 1]) {
      apiUrl = args[++i];
    } else if (args[i] === "--username" && args[i + 1]) {
      username = args[++i];
    } else if (args[i] === "--password" && args[i + 1]) {
      password = args[++i];
    }
  }

  if (!apiUrl || !username || !password) {
    console.error(
      "Usage: npx ts-node test_api.ts --api-url <URL> --username <USER> --password <PASS>"
    );
    process.exit(1);
  }

  return { apiUrl, username, password };
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
async function main(): Promise<void> {
  const { apiUrl, username, password } = parseArgs();

  console.log(`\n${"=".repeat(60)}`);
  console.log("  Cognito Auth System – End-to-End Tests");
  console.log(`${"=".repeat(60)}`);

  // 1. Login
  const tokens = await login(apiUrl, username, password);
  if (!tokens) {
    console.log("\nAborting – login failed.");
    process.exit(1);
  }

  let accessToken: string = tokens.AccessToken!;

  // 2. Refresh
  if (tokens.RefreshToken) {
    const newTokens = await refreshTokens(apiUrl, tokens.RefreshToken);
    if (newTokens?.AccessToken) {
      accessToken = newTokens.AccessToken;
      info("Using refreshed AccessToken for remaining tests");
    }
  }

  // 3. List events
  await testListEvents(apiUrl, accessToken);

  // 4. Create event
  const eventId = await testCreateEvent(apiUrl, accessToken);

  // 5. Get single event
  if (eventId) {
    await testGetEvent(apiUrl, accessToken, eventId);
  }

  // 6. Delete event
  if (eventId) {
    await testDeleteEvent(apiUrl, accessToken, eventId);
  }

  // 7. Unauthorized access (no token)
  await testUnauthorizedAccess(apiUrl);

  // 8. Invalid token
  await testInvalidToken(apiUrl);

  console.log(`\n${"=".repeat(60)}`);
  console.log("  Tests complete");
  console.log(`${"=".repeat(60)}\n`);
}

main();
